pennylane/transforms/compile.py

```diff
@@ -0,0 +1,166 @@
+            qml.RX(x, wires=0)
+            qml.CNOT(wires=[1, 0])
+            qml.RZ(-z, wires=2)
+            qml.RX(y, wires=2)
+            qml.PauliY(wires=2)
+            qml.CY(wires=[1, 2])
+            return qml.expval(qml.PauliZ(wires=0))
+
+    Visually, the original function looks like this:
+
+    >>> dev = qml.device('default.qubit', wires=[0, 1, 2])
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(0.2, 0.3, 0.4))
+     0: ──H──RX(0.4)──────╭X─────────RX(0.2)──╭X───────┤ ⟨Z⟩
+     1: ──H───────────╭X──╰C──────────────────╰C──╭CY──┤
+     2: ──H──RZ(0.4)──╰C───RZ(-0.4)──RX(0.3)───Y──╰CY──┤
+
+    We can compile it down to a smaller set of gates using the ``qml.compile``
+    transform.
+
+    >>> compiled_qfunc = qml.compile()(qfunc)
+    >>> compiled_qnode = qml.QNode(compiled_qfunc, dev)
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──H───RX(0.6)───────────────────┤ ⟨Z⟩
+     1: ──H──╭X─────────────────╭CY─────┤
+     2: ──H──╰C────────RX(0.3)──╰CY──Y──┤
+
+    You can change up the set of transforms by passing a custom ``pipeline`` to
+    ``qml.compile``. The pipeline is a list of transform functions. Furthermore,
+    you can specify a number of passes (repetitions of the pipeline), and a list
+    of gates into which the compiler will first attempt to decompose the
+    existing operations prior to applying any optimization transforms.
+
+    .. code-block:: python3
+
+        compiled_qfunc = qml.compile(
+            pipeline=[
+                qml.transforms.commute_controlled(direction="left"),
+                qml.transforms.merge_rotations(atol=1e-6),
+                qml.transforms.cancel_inverses
+            ],
+            basis_set=["CNOT", "RX", "RY", "RZ"],
+            num_passes=2
+        )(qfunc)
+
+        compiled_qnode = qml.QNode(compiled_qfunc, dev)
+
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──RZ(1.57)──RX(1.57)──RZ(1.57)───RX(0.6)───────────────────────────────────────────────────────────────────────┤ ⟨Z⟩
+     1: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╭X────────RZ(1.57)──────────────────────────────────────────╭C─────────────╭C──┤
+     2: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╰C────────RX(0.3)───RZ(1.57)──RY(3.14)──RZ(1.57)──RY(1.57)──╰X──RY(-1.57)──╰X──┤
+
+    """
+    # Ensure that everything in the pipeline is a valid qfunc or tape transform
+    if pipeline is None:
+        pipeline = default_pipeline
+    else:
+        for p in pipeline:
+            p_func = p.func if isinstance(p, partial) else p
+            if not isinstance(p_func, single_tape_transform) and not hasattr(p_func, "tape_fn"):
+                raise ValueError("Invalid transform function {p} passed to compile.")
+
+    if basis_set is not None:
+        for gate_name in basis_set:
+            if gate_name not in all_ops or not isinstance(gate_name, str):
+                raise ValueError("Invalid basis gate {gate_name} passed to compile.")
```

Comment:
I wonder if this check is necessary? e.g., imagine if a user is using a custom gate, or a gate from a plugin

Reply:
Both good points! Will remove the check. I suppose there's also no harm to the expansion process if someone passes an invalid gate name, it'll just get ignored.